src/lib/add-on/clients/add-on_auto.rb

```diff
@@ -1,328 +1,319 @@
 
-      Builtins.y2debug("ret=%1", @ret)
-      Builtins.y2milestone("add-on_auto finished")
-      Builtins.y2milestone("----------------------------------------")
+      media_url
+    end
+
+    # Expand url for given add_on
+    #
+    # @param [Hash] add_on
+    # @param [String] media_url
+    #
+    # @return [String] expanded url
+    def expand_url_for(add_on, media_url)
+      AddOnProduct.SetRepoUrlAlias(
+        Pkg.ExpandedUrl(media_url),
+        add_on.fetch("alias", ""),
+        add_on.fetch("name", "")
+      )
+    end
+
+    # Checks if should be retried to look for the source at given url
+    #
+    # @param [String] product
+    # @param [String] media_url
+    #
+    # @return [Boolean]
+    def retry_again?(product, media_url)
+      prompt = format(_("Make the add-on \"%s\" available via \"%s\"."), product, media_url)
+
+      Popup.ContinueCancel(prompt)
+    end
+
+    # Report an error about fail adding a product
+    #
+    # @param [String] product
+    # @param [String] media_url
+    def report_error_for(product, media_url)
+      # TRANSLATORS: The placeholders are for the product name and the URL.
+      # TRANSLATORS: a fallback string for undefined product name
+      product_name = product || _("<not_defined_name>")
+      error_msg = format(_("Failed to add product \"%s\" via\n%s"), product_name, media_url)
+
+      Report.Error(error_msg)
+    end
+
+    # Deletes given source and halt the system
+    #
+    # @param [Integer] source_id
+    def delete_and_halt(source_id)
+      log.warn("License not accepted, delete the repository and halt the system")
+
+      Pkg.SourceDelete(source_id)
+      SCR.Execute(path(".target.bash"), "/sbin/halt -f -n -p")
+    end
+
+    # Tries to confirm license if needed
+    #
+    # @param [Hash] add_on
+    # @param [Integer] source_id
+    #
+    # @return [Boolean] true if is not needed to confirm license or ir accepted; false otherwise
+    def accepted_license?(add_on, source_id)
+      return true unless add_on.fetch("confirm_license", false)
+
+      AddOnProduct.AcceptedLicenseAndInfoFile(source_id)
+    end
+
+    # Adjusts source attributes for given id
+    #
+    # At the moment to create source/repo through `Pkg.SourceCreate` is not possible to set attributes
+    # directly. In consequence, the creation must be completed making use of `Pkg.SourceEditSet`
+    #
+    # @see {https://github.com/yast/yast-pkg-bindings YaST Package Bindings}
+    #
+    # @param [Hash] add_on
+    # @param [Integer|Nil] source_id
+    def adjust_source_attributes(add_on, source_id)
+      repos = Pkg.SourceEditGet
+      repo_idx = repos.find_index { |repo| repo["SrcId"] == source_id }
+
+      return if repo_idx.nil?
+
+      repo = repos[repo_idx]
+      repo["name"] = preferred_name_for(add_on, repo)
+      repo["priority"] = add_on["priority"] if add_on.key?("priority")
+
+      log.info("Preferred name: #{repo["name"]}")
+
+      repos[repo_idx] = repo
```

Comment:
as ruby use reference and not copy, does it make sense this line?

Reply:
and if not, then all those index fun won't be needed and code can be much simplier.